drud/ddev
# Provide more useful message if removing of volume fails.

## The Issue

When volume removal fails, it might be tricky for less experienced users to find the reason. This PR adds information which containers are still using the volume.

## How This PR Solves The Issue

## Manual Testing Instructions
Create test project:
```
⇒ mkdir -p test-4635 && cd test-4635
⇒ ddev config --docroot=.  --project-type=php --project-name=test
⇒ ddev start
```

Confirm that only this project is using test-mariadb volume
```
⇒ docker ps --filter "volume=test-mariadb" -a
CONTAINER ID   IMAGE                                                COMMAND                  CREATED              STATUS                        PORTS                       NAMES
d586fab3af05   drud/ddev-dbserver-mariadb-10.4:v1.21.4-test-built   "/docker-entrypoint.…"   About a minute ago   Up About a minute (healthy)   127.0.0.1:32768->3306/tcp   ddev-test-db
```

Start two additional containers using that volume:
```
⇒ docker run -d --name test1 -v test-mariadb:/tmp hello-world:latest
⇒ docker run -d --name test2 -v test-mariadb:/tmp hello-world:latest
```

Confirm that all three are using test-mariadb volume now
```
⇒  docker ps --filter "volume=test-mariadb" -a                       
CONTAINER ID   IMAGE                                                COMMAND                  CREATED         STATUS                     PORTS                       NAMES
b807570dc649   hello-world:latest                                   "/hello"                 3 minutes ago   Exited (0) 3 minutes ago                               test2
391002c97206   hello-world:latest                                   "/hello"                 3 minutes ago   Exited (0) 3 minutes ago                               test1
d586fab3af05   drud/ddev-dbserver-mariadb-10.4:v1.21.4-test-built   "/docker-entrypoint.…"   5 minutes ago   Up 5 minutes (healthy)     127.0.0.1:32768->3306/tcp   ddev-test-db
```

Run `ddev delete -Oy` and get message 
```
could not remove volume test-mariadb: Docker volume 'test-mariadb' is in use by 2 containers and cannot be removed. Use 'docker rm -f test2 test1' to remove them 
```

Delete test2 `docker rm -f test2`

Run `ddev delete -Oy` and get message 
```
could not remove volume test-mariadb: Docker volume 'test-mariadb' is in use by a container 'test1' and cannot be removed. Use 'docker rm -f test1' to remove it
```

After deleting test1, you will be able to delete project without any problem

## Automated Testing Overview



## Related Issue Link(s)

## Release/Deployment Notes





<a href="https://gitpod.io/#https://github.com/drud/ddev/pull/4635"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
